src/main/java/org/icgc/argo/program_service/services/ProgramService.java

```diff
@@ -153,16 +161,25 @@ public ProgramEntity getProgram(@NonNull String name, boolean allowInactive) {
   }
 
   public ProgramEntity createWithSideEffect(
-      @NonNull Program program, Consumer<ProgramEntity> consumer) {
-    val programEntity = createProgram(program);
+      @NonNull Program program, Consumer<ProgramEntity> consumer, UUID dataCenterId) {
+    val programEntity = createProgram(program, dataCenterId);
     consumer.accept(programEntity);
     return programEntity;
   }
 
-  public ProgramEntity createProgram(@NonNull Program program)
+  public ProgramEntity createProgram(@NonNull Program program, UUID dataCenterId)
       throws DataIntegrityViolationException {
     val programEntity = programConverter.programToProgramEntity(program);
 
+    if (dataCenterId != null) {
+      val dataCenterEntity = dataCenterRepository.findById(dataCenterId);
+      if (!dataCenterEntity.isEmpty()) {
+        programEntity.setDataCenterId(dataCenterId);
+      } else {
+        throw new BadRequestException("DataCenterId '" + dataCenterId + "' not found");

```

Comment:
I think this is the wrong Error type to throw from a service. Since this is a service that could be called from any context, not exclusively from a web controller, throwing a BadRequest might not make sense to those other contexts. Is there another error type that is more relevant, that just communicates "Data not found" or similar? That error could be caught by the controller that calls it in that case return the BadRequestException.

This is not a practical issue for us since we won't have other uses of this service (at least not in our current plans), but its a better coding practice to throw relevant errors.

Reply:
Added RecordNotFoundException to handle scenario where the record doesn't exist in DB. This exception is handled as BadRequestException while throwing back to the caller of the endpoint.